### PR TITLE
[WIP] Add SSDlite architecture with MobileNetV3 backbones

### LINKED ISSUE #1422: please add other detectors
Please add other detectors like SSD.

### PR DESCRIPTION
**Pull request by datumbox**
Resolves #1422

### CODE REVIEW

File: torchvision/models/detection/ssdlite.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        for block in self.extra:
+            x = block(x)
+            output.append(x)
+
+        return OrderedDict([(str(i), v) for i, v in enumerate(output)])
+
+
+def _mobilenet_extractor(backbone_name: str, progress: bool, pretrained: bool, trainable_layers: int,
+                         norm_layer: Callable[..., nn.Module]):
+    backbone = mobilenet.__dict__[backbone_name](pretrained=pretrained, progress=progress,
```
Comment (datumbox) [inline torchvision/models/detection/ssdlite.py:133]: TODO: consider support of `width_mult`.

File: torchvision/models/detection/ssdlite.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    # Enable [-1, 1] rescaling and reduced tail if no pretrained backbone is selected
+    rescaling = reduce_tail = not pretrained_backbone
+
+    if norm_layer is None:
+        norm_layer = partial(nn.BatchNorm2d, eps=0.001, momentum=0.03)
+
+    backbone = _mobilenet_extractor("mobilenet_v3_large", progress, pretrained_backbone, trainable_backbone_layers,
+                                    norm_layer, rescaling, _reduced_tail=reduce_tail, _width_mult=1.0)
+
+    size = (320, 320)
```
Comment (fmassa) [inline torchvision/models/detection/ssdlite.py:204]: This means that the size is hard-coded and even if the user passes a different `size` `**kwargs` in the constructor it won't be used?

What about doing something like
```python
size = kwargs.get("size", (320, 320))
```
instead, so that the users can potentially customize the input size if they wish?

**Review by fmassa — COMMENTED** [review]
Looks great, thanks a lot Vasilis!

I have a couple of comments, let me know what you think

File: torchvision/models/detection/ssdlite.py (hunk 2 above)
Comment (datumbox) [reply torchvision/models/detection/ssdlite.py:204]: I chose to hardcode it because this is the `ssdlite320` model which uses a fixed 320x320 size. The input size is much less flexible on SSD models comparing to FasterRCNN because they make a few strong assumptions about the input. 

If someone wants to use a different size, it would be simpler to just create the backbone, configure the DefaultBoxGenerator and then initialize directly the `SSD` with the config of their choice. Overall I felt that this approach would be simpler than trying to offer an API that tries to cover all user needs.

File: torchvision/models/detection/ssdlite.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    defaults = {
+        "score_thresh": 0.001,
+        "nms_thresh": 0.55,
+        "detections_per_img": 300,
+        "topk_candidates": 300,
+    }
+    kwargs = {**defaults, **kwargs}
+    model = SSD(backbone, anchor_generator, size, num_classes,
+                head=SSDLiteHead(out_channels, num_anchors, num_classes, norm_layer),
+                image_mean=[0., 0., 0.], image_std=[1., 1., 1.], **kwargs)
```
[the remark this answers is not in the document]
Comment (datumbox) [reply torchvision/models/detection/ssdlite.py:219]: Correct. Here I'm trying to be as close to the canonical implementation as possible and that helped me close the gap in the accuracy.

You are right to say that a pretrained backbone would need different mean/std. Thankfully because our setup is to train end-to-end and use extensive BN, the backbone adapts to the different input fairly quickly even when one uses a pre-trained backbone. In the end, as I trained it for quite a few epochs, it was better to start from random weights which led to a better result (this is a common finding in similar setups).

Though indeed overall for the API it might be simpler to disable passing a pretrained backbone, this means that the API for SSDlite will be different from any other model. It will also create issues with our training scripts that expect to be able to pass this parameter. I think what I will do to address better this remark is make the mean/std configurable.

**Review by fmassa — APPROVED** [review]
Thanks for the answers Vasilis!

Chatted offline, let's get this PR merged and then follow up with creating a few issues to investigate some of the points that I brought.

File: torchvision/models/detection/ssdlite.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        width_mult = kwargs.pop('_width_mult', 1.0)
+
+        assert not backbone[c4_pos].use_res_connect
+        self.features = nn.Sequential(
+            nn.Sequential(*backbone[:c4_pos], backbone[c4_pos].block[0]),  # from start until C4 expansion layer
+            nn.Sequential(backbone[c4_pos].block[1:], *backbone[c4_pos + 1:]),  # from C4 depthwise until end
+        )
+
+        get_depth = lambda d: max(min_depth, int(d * width_mult))  # noqa: E731
+        extra = nn.ModuleList([
```
Comment (evekeen) [inline torchvision/models/detection/ssdlite.py:111]: @datumbox could you please help me figure it out - I cannot find the info about these extra layers in the papers. Where did you get them from?
I'm trying to create a modification for this model and struggle to understand it - any help would be appreciated!
I want to reduce the number of encoder layers to make feature maps detect small objects.

Comment (evekeen) [reply torchvision/models/detection/ssdlite.py:111]: @datumbox Thank you for the quick reply! It's very helpful

Comment (evekeen) [reply torchvision/models/detection/ssdlite.py:111]: @datumbox in 6.3 of MobileNet3 paper, I only see the info on connecting C4 and C5 layers to the SSD head. There is nothing on these extra layers there.

Comment (datumbox) [reply torchvision/models/detection/ssdlite.py:111]: Have you checked the reference code I sent? This comes from their official repo.

Comment (evekeen) [reply torchvision/models/detection/ssdlite.py:111]: Yes, I see that in the TensorFlow implementation.

Comment (datumbox) [reply torchvision/models/detection/ssdlite.py:111]: Sorry, it's been quite sometime since I wrote the implementation. I think you will need to dig into the original research repo to get the details.
